File: DSApython/Algorithm/Sorting/mergeSort.py
def merge(arr1,arr2,into):
    i1,i2,a = 0,0,0
    while a<len(into):
        while i1 < len(arr1) and (i2>=len(arr2) or arr1[i1]<arr2[i2]):
            into[a] = arr1[i1]
            a+=1
            i1+=1
        while i2 < len(arr2) and (i1>=len(arr1) or arr2[i2]<arr1[i1]):
            into[a] = arr2[i2]
            a+=1
            i2+=1

def mergeSort(arr):
    left = 0
    right = len(arr)-1
    if left>=right:
        return
    
    mid = (right-left)>>1

    arr1 = [arr[i] for i in range(left,mid+1)]
    mergeSort(arr1)

    arr2 = [arr[i] for i in range(mid+1,right+1)]
    mergeSort(arr2)

    merge(arr1,arr2,arr)

arr = [10,9,8,7,6,5,4,3,2,1]
mergeSort(arr)
print(arr)

File: DSAcpp/Algorithm/Sortings/mergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void printArr(const vector<int> &arr){
    for(int i:arr) cout << i << ' ';
    cout << '\n';
}

vector<int> copy(const vector<int> & arr,int l, int r){
    vector<int> copyArr((r-l+1));
    for(int i=l;i<=r;i++) copyArr[i-l] = arr[i];
    return copyArr;
}

void marge(const vector<int> &arr1, const vector<int> &arr2, vector<int> &into){
    int i1=0,i2=0,a=0;
    while(a<into.size()){
        while(i1<arr1.size() && (arr1[i1]<arr2[i2] || i2==arr2.size())){
            into[a++] = arr1[i1++];
        }
        while(i2<arr2.size() && (arr2[i2]<arr1[i1] || i1==arr1.size())){
            into[a++] = arr2[i2++];
        }
    }
}

void mergeSort(vector<int> &arr){
    int left = 0, right=arr.size()-1;
    if (left>=right) return;
    
    int mid = (right-left)>>1;

    vector<int> arr1 = copy(arr,left,mid);
    mergeSort(arr1);

    vector<int> arr2 = copy(arr,mid+1,right);
    mergeSort(arr2);

    marge(arr1,arr2,arr);
}

int main(){
    vector<int> arr = {10,9,8,7,6,5,4,3,2,1};
    mergeSort(arr);
    printArr(arr);
}